### ISSUE TITLE
"Bad gateway" error when requesting an analysis

### ISSUE THREAD

**Issue by mcrot** [ContactEngineering/topobank#671]
![grafik](https://user-images.githubusercontent.com/9113541/119992400-cbe4ca80-bfca-11eb-85f5-9c51a6a1a9ec.png)

This happened while the requested analysis was not ready yet. Afterwards I had a look and it seems to have been started,
since it is marked in the database as "started".

**Comment by mcrot** [comment]
Interestingly, in the database the state is "started", but from celery's point of view, it's "pending":
```python
>>> from topobank.taskapp.tasks import perform_analysis
>>> ar=perform_analysis.AsyncResult("d4700200-865a-41d5-8a19-f59d91fee92c")
>>> ar.
ar.TimeoutError(         ar.children              ar.graph                 ar.maybe_throw(          ar.retries               ar.then(
ar.app                   ar.collect(              ar.id                    ar.name                  ar.revoke(               ar.throw(
ar.args                  ar.date_done             ar.ignored               ar.on_ready(             ar.state                 ar.traceback
ar.as_list(              ar.failed(               ar.info                  ar.parent                ar.status                ar.wait(
ar.as_tuple(             ar.forget(               ar.iterdeps(             ar.queue                 ar.successful(           ar.worker
ar.backend               ar.get(                  ar.kwargs                ar.ready(                ar.supports_native_join  
ar.build_graph(          ar.get_leaf(             ar.maybe_reraise(        ar.result                ar.task_id               
>>> ar.status
'PENDING'
>>> ar.state
'PENDING'
```
I've restarted the analysis:
```python
$ docker-compose -f production.yml run --rm django python manage.py trigger_analyses a104306
Starting topobank_rabbitmq_1 ... done
Starting topobank_postgres_1  ... done
Starting topobank_memcached_1 ... done
PostgreSQL is available
INFO 2021-05-28 15:01:29,059 utils 1 139927379146112 Renewing analysis 104306 for 2 users, function Contact Mechanics, subject type topography, subject id 4480 .. kwargs: {'substrate_str': None, 'hardness': None, 'nsteps': 10, 'pressures': None, 'maxiter': 100}
INFO 2021-05-28 15:01:29,491 signals 1 139927379146112 Deleting analyses/104306/result-step-0.nc..
INFO 2021-05-28 15:01:29,504 signals 1 139927379146112 Deleting analyses/104306/result-step-1.nc..
INFO 2021-05-28 15:01:29,546 signals 1 139927379146112 Deleting analyses/104306/result-step-2.nc..
INFO 2021-05-28 15:01:29,566 signals 1 139927379146112 Deleting analyses/104306/result-step-3.nc..
Renewed analysis 'Task d4700200-865a-41d5-8a19-f59d91fee92c with state started'.
```
Now the state is "pending" in database and UI. There are currently a lot of load on the workers, so the job couldn't be taken so far.

**Comment by mcrot** [comment]
Now the job changed to "started" in database. The UI reports "Failure" (report by API request to celery). 
The database is still at "started".

When looking into the backend, the job as indeed failed because it was killed:
```python
>>> from topobank.taskapp.tasks import perform_analysis
>>> ar=perform_analysis.AsyncResult('5b7cd217-041a-4531-9971-f7586b8e462c')
>>> ar.state
'FAILURE'
>>> ar.
ar.TimeoutError(         ar.children              ar.graph                 ar.maybe_throw(          ar.retries               ar.then(
ar.app                   ar.collect(              ar.id                    ar.name                  ar.revoke(               ar.throw(
ar.args                  ar.date_done             ar.ignored               ar.on_ready(             ar.state                 ar.traceback
ar.as_list(              ar.failed(               ar.info                  ar.parent                ar.status                ar.wait(
ar.as_tuple(             ar.forget(               ar.iterdeps(             ar.queue                 ar.successful(           ar.worker
ar.backend               ar.get(                  ar.kwargs                ar.ready(                ar.supports_native_join  
ar.build_graph(          ar.get_leaf(             ar.maybe_reraise(        ar.result                ar.task_id               
>>> ar.kwargs
>>> ar.info
WorkerLostError('Worker exited prematurely: signal 9 (SIGKILL).')
```
It seems this is the cause for #672 .

So the job was killed (too much resources?) but this was not saved in the database backend.

So there are multiple questions here:
- Why has the job been killed?
- Why is there no database entry about the failure? The celery backend knows about it.
- Why was "Bad gateway" shown to the user instead of a red progress bar saying "Failure"?

**Comment by mcrot** [comment]
Hi @AntoineSIMTEK , your job has been killed in the background. Or more precisely, the worker process was killed on which your analysis was running. 

- Does this analysis take much memory?
- Significantly more than for the other measurements in this surface?
- Could you estimate how much memory is needed? The measurement has 2048 x 2048 points.

**Comment by AntoineSIMTEK** [comment]
In the same surface there are topographies of the same size. 

See for example rough_sphere_pas_robb_fig1.txt

**Comment by AntoineSIMTEK** [comment]
I estimate it needs between 1 and 2 Gb

**Comment by mcrot** [comment]
Ah, the "Bad gateway" message appears after some time waiting for all three progress bars to finish (for all three measurements).
This is not an expected behavior, I expected now two bars would be green (complete), because the corresponding jobs were complete. 
So I have to check several things in this issue.

**Comment by AntoineSIMTEK** [comment]
I don't think it needs more resources. But I think the algorithm could fail.

The rough sphere behaves a bit like a rough surface, and the guesses for the applied forces are made based on the roughness. 

It can be that the guesses are so bad that the algorithm has no success.

**Comment by mcrot** [comment]
Yes, again this one fails. And again it seems that it has been killed.

So what I can do is to make it fail more gracefully. I'm not sure what I can do to make this job succeed.
Have you tried to do the calculation "manually" in a notebook? Does it behave differently?

**Comment by AntoineSIMTEK** [comment]
I guess that if I was able to change the list normal forces I could choose reasonable values that wouldn't fail.

**Comment by AntoineSIMTEK** [comment]
But I will check if it works calling directly the backend

**Comment by AntoineSIMTEK** [comment]
If I choose reasonable forces, the backend works. 

![image](https://user-images.githubusercontent.com/40853209/120000171-d3a86d00-bfd2-11eb-8261-ff233fdc682c.png)


even with the high yellow contact area. 

The pentol was 1e-6, and I expect the heuristic backend to have chosen 1e-4

**Comment by AntoineSIMTEK** [comment]
When the force is to high I need 200 iter and it looks like this: 
![Uploading image.png…]()

**Comment by mcrot** [comment]
So you also experienced extensive memory usage if you don't use reasonable forces? Would that explain that the job has been killed in TopoBank?

**Comment by mcrot** [comment]
However this analysis also eats up all the memory an my development machine with 32 GB. 
Finally the job is killed by the OS. I'll shift this to the next version, then we plan to shift the workers to a machine with more RAM.

**Comment by pastewka** [comment]
Closed by #965
